### PR TITLE
Add list-outputs

### PR DESCRIPTION
**Pull request by bshifaw**
Add list-outputs command
Add unit and integrations tests

### CODE REVIEW

File: tests/integration/utility_test_functions.py [hunk 1]
```diff
@@ -9,10 +9,11 @@
 from cromshell.utilities import cromshellconfig
 
 
-def run_cromshell_command(command: list, exit_code: int):
+def run_cromshell_command(command: list, exit_code: int, options: list = None):
```
Comment (SHuang-Broad) [inline tests/integration/utility_test_functions.py:12]: Let's document what these parameters are.

File: tests/integration/utility_test_functions.py [hunk 2]
```diff
@@ -24,12 +25,18 @@ def run_cromshell_command(command: list, exit_code: int):
     :return: results from execution
     """
 
+    command_with_options = command.copy()
+    if options:
+        for item in options:
+            command_with_options.insert(1, item)
```
Comment (SHuang-Broad) [inline tests/integration/utility_test_functions.py:31]: what if the options are meant to be in order? This seems to reverse that.

File: tests/unit/test_list_outputs.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+import pytest
+
+from cromshell.list_outputs import command as list_outputs_command
+
+
+class TestListOutputs:
+    """Test the execution list-outputs command functions"""
+
+    # def get_workflow_level_outputs(config) -> dict:
```
Comment (SHuang-Broad) [inline tests/unit/test_list_outputs.py:11]: is this a self-reminder? If so, let's be a bit explicit in the comments.

File: src/cromshell/list_outputs/command.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        indent (bool): Whether to indent the output"""
+
+    i = "\t" if indent else ""
+
+    if isinstance(output_value, str):
+        if is_path_or_url_like(output_value):
+            print(f"{i}{output_name}: {output_value}")
+
+
+def is_path_or_url_like(in_string: str) -> bool:
```
Comment (SHuang-Broad) [inline src/cromshell/list_outputs/command.py:186]: I'm thinking about future-proofing us a little here: how about we classify the `in_string`? 
I mean classify it as a local regular path, GCS path, or Azure path, or not a path at all.

File: src/cromshell/list_outputs/command.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        timeout=config.requests_connect_timeout,
+        verify=config.requests_verify_certs,
+        headers=http_utils.generate_headers(config),
+    )
+
+    if requests_out.ok:
+        return requests_out.json()
+    else:
+        http_utils.check_http_request_status_code(
+            short_error_message="Failed to retrieve workflow outputs.",
```
Comment (SHuang-Broad) [inline src/cromshell/list_outputs/command.py:80]: It'd be better if we also include the ID of this workflow that fails.

File: src/cromshell/list_outputs/command.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            workflow_id=workflow_id, cromshell_config=config
+        )
+
+        if not detailed:
+            if json_summary:
+                io_utils.pretty_print_json(
+                    format_json=get_workflow_level_outputs(config).get("outputs")
+                )
+            else:
+                print_file_like_value_in_dict(
```
Comment (SHuang-Broad) [inline src/cromshell/list_outputs/command.py:49]: Does this handle `Array[File]` outputs?

File: src/cromshell/list_outputs/command.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        if "subWorkflowMetadata" in calls_metadata[call][0]:
+            output_metadata[call] = []
+            for scatter in calls_metadata[call]:
+                output_metadata[call].append(
+                    filer_outputs_from_workflow_metadata(scatter["subWorkflowMetadata"])
+                )
+        else:
+            output_metadata[call] = []
+            for index in index_list:
+                output_metadata[call].append(index.get(extract_task_key))
```
Comment (SHuang-Broad) [inline src/cromshell/list_outputs/command.py:132]: what if a task has no output? In theory it is possible. Then the `get` call will return `None`. Does the downstream handle `None` correctly? I think it's better if we handle that here.

**Review by SHuang-Broad — CHANGES_REQUESTED** [review]
Let's also make sure #249 is reflected here.

File: tests/unit/test_list_outputs.py (hunk 3 above)
Comment (bshifaw) [reply tests/unit/test_list_outputs.py:11]: oops, that should have been deleted

File: tests/integration/utility_test_functions.py (hunk 1 above)
Comment (bshifaw) [reply tests/integration/utility_test_functions.py:12]: done, also changed name to subcommand_options

File: src/cromshell/list_outputs/command.py (hunk 6 above)
Comment (bshifaw) [reply src/cromshell/list_outputs/command.py:49]: It should handle `output_name: File` and `output_name: [File, File,..]`

File: src/cromshell/list_outputs/command.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    LOGGER.info("list-outputs")
+
+    return_code = 0
+
+    for workflow_id in workflow_ids:
+        command_setup_utils.resolve_workflow_id_and_server(
+            workflow_id=workflow_id, cromshell_config=config
+        )
+
+        if not detailed:
```
[the remark this answers is not in the document]
Comment (bshifaw) [reply src/cromshell/list_outputs/command.py:43]: done

File: src/cromshell/list_outputs/command.py (hunk 7 above)
Comment (bshifaw) [reply src/cromshell/list_outputs/command.py:136]: good catch, done

File: src/cromshell/list_outputs/command.py (hunk 5 above)
Comment (bshifaw) [reply src/cromshell/list_outputs/command.py]: done

File: src/cromshell/utilities/command_setup_utils.py [hunk 9]
```diff
@@ -17,5 +17,17 @@ def resolve_workflow_id_and_server(workflow_id: str, cromshell_config) -> str:
     http_utils.set_and_check_cromwell_server(
         config=cromshell_config, workflow_id=resolved_workflow_id
     )
+    set_workflow_id(workflow_id=resolved_workflow_id, cromshell_config=cromshell_config)
 
     return resolved_workflow_id
+
+
+def set_workflow_id(workflow_id: str, cromshell_config) -> None:
```
Comment (SHuang-Broad) [inline src/cromshell/utilities/command_setup_utils.py:25]: I'm concerned about this.
What about commands that support multiple workflow IDs?

Comment (SHuang-Broad) [reply src/cromshell/utilities/command_setup_utils.py:25]: resolved during code review meeting.

**Pull request merged**